File: leet_Find Two Non-overlapping Sub-arrays Each With Target Sum.py
"""
Intution:
we are storing best possible subarray length in right[i] and summing up
"""
class Solution:
    def minSumOfLengths(self, arr: List[int], target: int) -> int:
class Solution:
    def minSumOfLengths(self, arr: List[int], target: int) -> int:
        ha= {0:-1}
        first = [float('inf')]*(len(arr)+1)
        curr = 0
        res = float('inf')
        for i,n in enumerate(arr):
            curr+=n
            if curr-target in ha:
                ind = ha[curr-target]
                first[i+1] = min(i-ind,first[i])
                res = min(res,first[ind+1]+(i-ind))
            else:
                first[i+1] = first[i]
            ha[curr] = i
        return res if res!=float('inf') else -1
        
                
        # right=[float('inf')]*len(arr)
        # l,r=0,0
        # cur = 0
        # for r in range(len(arr)):
        #     cur+=arr[r]
        #     if cur == target:
        #         right[r] = r-l+1
        #         cur-=arr[l]
        #         l+=1
        #     elif cur>target:
        #         while cur>target:
        #             cur-=arr[l]
        #             l+=1
        #         if cur==target:
        #             right[r] = r-l+1
        #             cur-=arr[l]
        #             l+=1
            # r+=1
        # # print(right)
        # left=[float('inf')]*len(arr)
        # curMin = float('inf')
        # for i in range(len(arr)-1,-1,-1):
        #     if right[i]!=float('inf'):
        #         left[i-right[i]+1] = min(right[i],left[i+1] if i!=len(arr)-1 else float('inf'))
        #     if i!=len(arr)-1:
        #         left[i]=min(left[i],left[i+1])
        # res=float('inf')
        # for i in range(1,len(right)):
        #     if i-right[i]>=0 and right[i]!=float('inf') and right[i-right[i]]!=float('inf'):
        #         res = min(res,right[i]+right[i-right[i]])
        #     right[i]=min(right[i],right[i-1])
        # # for i in range(len(arr)-1):
        # #     res = min(res,right[i]+left[i+1])
        # return res if res!=float('inf') else -1

File: LearnJava/leet_find_two_non_overlapping_subarrays_each_with_target_sum.java
class Solution {
    public int minSumOfLengths1(int[] arr, int target) {
        int res = Integer.MAX_VALUE;
        int[] pref = new int[arr.length+1];
        pref[0] = Integer.MAX_VALUE;
        Map<Integer,Integer> ha  = new HashMap<>();
        ha.put(0,0);
        int sm=0;
        for(int i=0;i<arr.length;i++){
            sm+=arr[i];
            if(ha.containsKey(sm-target)){
                int ind = ha.get(sm-target);
                res = pref[ind]!=Integer.MAX_VALUE?Math.min(res,i-ind+1+pref[ind]):res;
                pref[i+1] = Math.min(i-ind+1,pref[i]);
                // System.out.println(res);
            }
            else{
                pref[i+1] = pref[i];
            }
            ha.put(sm,i+1);
        }
        // System.out.println(ha);
        return res!=Integer.MAX_VALUE?res:-1;
    }
   public int minSumOfLengths(int[] arr, int target) {
        int res = Integer.MAX_VALUE;
        int[] pref = new int[arr.length+1];
        Arrays.fill(pref,Integer.MAX_VALUE);
        int sm=0,left=0,right,tmp;
        for(right=0;right<arr.length;right++){
            sm+=arr[right];
            while(sm>target)sm-=arr[left++];
            
            if(sm == target){
                // System.out.print(right);
                // System.out.print(left);
                // System.out.println(Arrays.toString(pref));
                tmp = right-left+1;
                if(pref[left] != Integer.MAX_VALUE){
                    res = Math.min(res,tmp+pref[left]);    
                }
                pref[right+1] = Math.min(pref[right],tmp);
            }
            else{
                pref[right+1] = pref[right];
            }
        }
        // System.out.println(ha);
        return res!=Integer.MAX_VALUE?res:-1;
    }
}